crates/wascc-provider/src/lib.rs

```diff
@@ -253,20 +331,61 @@ impl kubelet::state::AsyncDrop for PodState {
             }
         }
         {
-            let mut handles = self.shared.handles.write().await;
+            let mut handles = provider_state.s.handles.write().await;
             handles.remove(&self.key);
         }
     }
 }
 
+impl kubelet::state::common::GenericProvider for WasccProvider {
+    type PodState = PodState;
+    type ProviderState = ProviderState;
+    type RunState = crate::states::starting::Starting;
+
+    fn validate_pod_runnable(pod: &kubelet::pod::Pod) -> anyhow::Result<()> {
+        if !pod.init_containers().is_empty() {
+            return Err(anyhow::anyhow!(
+                "Cannot run {}: spec specifies init containers which are not supported on wasCC",
+                pod.name()
+            ));
+        }
+        Ok(())
+    }
+
+    fn validate_container_runnable(
+        container: &kubelet::container::Container,
+    ) -> anyhow::Result<()> {
+        if has_args(container) {
+            return Err(anyhow::anyhow!(
+                "Cannot run {}: spec specifies container args which are not supported on wasCC",
+                container.name()
+            ));
+        }
+        if let Some(image) = container.image()? {
+            if image.whole().starts_with("k8s.gcr.io/kube-proxy") {
+                return Err(anyhow::anyhow!("Cannot run kube-proxy"));
+            }
+        }
+
+        Ok(())
+    }
+}
+
 #[async_trait]
 impl Provider for WasccProvider {
-    type InitialState = Registered;
-    type TerminatedState = Terminated;
+    type InitialState = Registered<Self>;
```

Comment:
This would be completely non-blocking if we move forward with this PR, but there seems to be quite a few extra steps needed to use the generic states. Although this specific part falls firmly within nit territory, it seems a little awkward to be making the initial state require `Self`. But beyond that nit, there are a whole bunch of extra traits you have to add in order to use the generic state things. Like @kflansburg mentioned, I am ok with this for now seeing as it is optional, but I think we should keep iterating on this to make it a bit less effort.

To be clear: I don't have another solution in mind right now, and I'd rather have something reusable like this rather than nothing. Just expressing initial impressions based on how I am seeing this implemented in a provider here